firebase-firestore/src/test/java/com/google/firebase/firestore/model/FieldIndexMatcherTest.java

```diff
@@ -0,0 +1,478 @@
+  }
+
+  @Test
+  public void withNotEquals() {
+    Query q = query("collId").filter(filter("a", "!=", 1));
+    validateServesTarget(q, "a", FieldIndex.Segment.Kind.ORDERED);
+
+    q = query("collId").filter(filter("a", "!=", 1)).orderBy(orderBy("a")).orderBy(orderBy("b"));
+    validateServesTarget(
+        q, "a", FieldIndex.Segment.Kind.ORDERED, "b", FieldIndex.Segment.Kind.ORDERED);
+  }
+
+  @Test
+  public void withMultipleFilters() {
+    Query queriesMultipleFilters =
+        query("collId").filter(filter("a", "==", "a")).filter(filter("b", ">", "b"));
+    validateServesTarget(queriesMultipleFilters, "a", FieldIndex.Segment.Kind.ORDERED);
+    validateServesTarget(
+        queriesMultipleFilters,
+        "a",
+        FieldIndex.Segment.Kind.ORDERED,
+        "b",
+        FieldIndex.Segment.Kind.ORDERED);
+  }
+
+  @Test
+  public void multipleFiltersRequireMatchingPrefix() {
+    Query queriesMultipleFilters =
+        query("collId").filter(filter("a", "==", "a")).filter(filter("b", ">", "b"));
+
+    validatePartiallyServesTarget(
+        queriesMultipleFilters,
+        /* expectedSegmentCount= */ 1,
+        "b",
+        FieldIndex.Segment.Kind.ORDERED);
+    validateDoesNotServeTarget(
+        queriesMultipleFilters,
+        "c",
+        FieldIndex.Segment.Kind.ORDERED,
+        "a",
+        FieldIndex.Segment.Kind.ORDERED);
+  }
+
+  @Test
+  public void withMultipleFiltersAndOrderBy() {
+    Query queriesMultipleFilters =
+        query("collId")
+            .filter(filter("a1", "==", "a"))
+            .filter(filter("a2", ">", "b"))
+            .orderBy(orderBy("a2", "asc"));
+    validateServesTarget(
+        queriesMultipleFilters,
+        "a1",
+        FieldIndex.Segment.Kind.ORDERED,
+        "a2",
+        FieldIndex.Segment.Kind.ORDERED);
+  }
+
+  @Test
+  public void withMultipleInequalities() {
+    Query q =
+        query("collId")
+            .filter(filter("a", ">=", 1))
+            .filter(filter("a", "==", 5))
+            .filter(filter("a", "<=", 10));
+    validateServesTarget(q, "a", FieldIndex.Segment.Kind.ORDERED);
+  }
+
+  @Test
+  public void withMultipleNotIn() {
+    Query q =
+        query("collId")
+            .filter(filter("a", "not-in", Arrays.asList(1, 2, 3)))
+            .filter(filter("a", ">=", 2));
+    validateServesTarget(q, "a", FieldIndex.Segment.Kind.ORDERED);
+  }
+
+  @Test
+  public void withMultipleOrderBys() {
+    Query q =
+        query("collId")
+            .orderBy(orderBy("fff"))
+            .orderBy(orderBy("bar", "desc"))
+            .orderBy(orderBy("__name__"));
+    validateServesTarget(
+        q,
+        "fff",
+        FieldIndex.Segment.Kind.ORDERED,
+        "bar",
+        FieldIndex.Segment.Kind.ORDERED,
+        "__name__",
+        FieldIndex.Segment.Kind.ORDERED);
+    // The field index can be used regardless of the order of segments since we only use the orderBy
```

Comment:
I see. 

So `orderBy` here serves the purpose of `fieldExists`. And this is different from using no index at all, because that would include documents without these fields as well.

Then what about an index of `(fff, xxx, bar, __name__)`, this would match with `fff`, but would only include documents having `xxx`, which seems to be wrong?

What about `(bar, __name__)`, this would be useful, but we need to do postfiltering to remove documents without `fff`. Is this understanding correct?

Reply:
Your understanding is correct. I think we can keep most of our logic, but must change how we handle "index subsets". Given your example, I don't think we can actually serve a documents from an index that indexes more fields than the target has. This further simplifies this PR as we can now return a simple boolean to indicate whether an Index matches a target.

Thanks for catching this flaw! I think all of this should be caught by the conformance test suite when we put everything together, but it is always better to catch things early.

Reply:
Maybe I missed it, can you add a test for `(fff, xxx, bar, __name__)` to make sure this returns false?

Also, I don't think we should resolve it here, but index with a super set fields cannot serve a target seems a bit unfortunate. Can we consider indexing documents with absent fields?

Reply:
Added one more test.

We might have to add an additional index if we decide to index documents with missing fields since we will no longer be able to guarantee `O(result set)` performance even for queries that do have their own index.

Reply:
Yeah, we need to decide what to do here. But this PR LGTM.